apps/oxlint/src/output_formatter/stylish.rs

```diff
@@ -0,0 +1,146 @@
+use std::io::Write;
+
+use oxc_diagnostics::{
+    reporter::{DiagnosticReporter, Info},
+    Error, Severity,
+};
+use rustc_hash::FxHashMap;
+
+use crate::output_formatter::InternalFormatter;
+
+#[derive(Debug, Default)]
+pub struct StylishOutputFormatter;
+
+impl InternalFormatter for StylishOutputFormatter {
+    fn all_rules(&mut self, writer: &mut dyn Write) {
+        writeln!(writer, "flag --rules with flag --format=stylish is not allowed").unwrap();
+    }
+
+    fn get_diagnostic_reporter(&self) -> Box<dyn DiagnosticReporter> {
+        Box::new(StylishReporter::default())
+    }
+}
+
+#[derive(Default)]
+struct StylishReporter {
+    diagnostics: Vec<Error>,
+}
+
+impl DiagnosticReporter for StylishReporter {
+    fn finish(&mut self) -> Option<String> {
+        Some(format_stylish(&self.diagnostics))
+    }
+
+    fn render_error(&mut self, error: Error) -> Option<String> {
+        self.diagnostics.push(error);
+        None
+    }
+}
+
+fn format_stylish(diagnostics: &[Error]) -> String {
+    if diagnostics.is_empty() {
+        return String::new();
+    }
+
+    let mut output = String::new();
+    let mut total_errors = 0;
+    let mut total_warnings = 0;
+
+    let mut grouped: FxHashMap<String, Vec<&Error>> = FxHashMap::default();
+    let mut sorted = diagnostics.iter().collect::<Vec<_>>();
+
+    sorted.sort_by_key(|diagnostic| Info::new(diagnostic).line);
+
+    for diagnostic in sorted {
+        let info = Info::new(diagnostic);
+        grouped.entry(info.filename).or_default().push(diagnostic);
+    }
+
+    for diagnostics in grouped.values() {
+        let diagnostic = diagnostics[0];
+        let info = Info::new(diagnostic);
+        let filename = info.filename;
+        let filename = if let Some(path) =
+            std::env::current_dir().ok().and_then(|d| d.join(&filename).canonicalize().ok())
```

Comment:
`canonicalize` adds under widows a  `\\?\` prefix to it.
Is there a problem using `info.filename` without it? 
Would like to avoid `current_dir` and `canonicalize` :)

Reply:
@Sysix a note from my PR description:

> Please note that this is my first time working with Rust and my knowledge is very limited. I'm unlikely to understand best-practice or best-pattern references outside of what clippy/cargo lint has already had me change. If this needs modification, please help me out by making code suggestions that can be merged to this PR.

I'm also not a Windows user, so I'd have no valuable insight into that question. Please do make a code suggestion

Reply:
I would like to see this code block removed:

```rust
        let filename = if let Some(path) =
            std::env::current_dir().ok().and_then(|d| d.join(&filename).canonicalize().ok())
        {
            path.display().to_string()
        } else {
            filename
        };
```

`let filename = info.filename;` should be just fine

Reply:
> let filename = info.filename; should be just fine.

When I was testing the output using `just watch` the filename was relative to the current working directory, and would print just `test.js` using `let filename = info.filename;`. ESLint v9 always prints the full file path, and so I was attempting to get to 1:1 with that. Personally, I'd very much prefer the full file path. Maybe I did something else incorrectly in conjunction, but the solution as it is was the only way I could get the reporter to get to the full path. 

> PS: thanks you for the work 🫶

very welcome. this is an impressive project, happy to try and make a modest contribution.

Reply:
Got it! No work on your side. Great job 🥳 
You can resolve the discussion, I do not have the rights for this :)